docs/operations.rst

```diff
@@ -14,42 +14,169 @@ Provisioning
-parameter when creating monitor or data Node. Any auth method supported by
-PostgreSQL could be used here. Please refer to `PostgreSQL pg_hba documentation`__
-for available options.
+  - on the monitor, the internal pg_auto_failover database schema might have
+    changed and needs to be upgraded to its new definition, porting the
+    existing data over. The pg_auto_failover database contains the
+    registration of every node in the system and their current state.
 
-__ https://www.postgresql.org/docs/current/auth-pg-hba-conf.html
+	It is not possible to trigger a failover during the monitor update.
+	Postgres operations on the Postgres nodes continue normally.
 
-Security for following connections should be considered when setting up
-`.pgpass` file.
+	During the restart of the monitor, the other nodes might have trouble
+	connecting. The ``pg_autoctl`` command is designed to retry connecting
+	to the monitor and handle errors gracefully. It also checks of the
+	expected Postgres replication connections are still being made, and when
+	that's the case, nodes will continue operating normally during the
+	monitor ugprade.
 
-  1. health check connection from monitor for `autoctl` user to both `postgres` and `pg_auto_failover` databases.
-  2. connections for `pg_autoctl` command from data nodes to monitor for `autoctl_node` user.
-  3. replication connections from secondary to primary data nodes for `replication` user.
-     Notice that primary and secondary nodes change during failover. Thus this setting
-     should be done on both primary and secondary nodes.
-  4. settings need to be updated after a new node is added.
+  - on the Postgres nodes, the ``pg_autoctl`` command connects to the
+    monitor every once in a while (every second by default), and then calls
+    the ``node_active`` protocol, a stored procedure in the monitor databases.
 
-See `PostgreSQL documentation`__ on setting up `.pgpass` file.
+	When upgrading the monitor, it may happen that this procedure has a new
+	API contract. In that case queries from the running ``pg_autoctl`` are
+	going to fail. We need to upgrade the client program so that we have the
+	code that matches the monitor extension's version.
 
-__ https://www.postgresql.org/docs/current/libpq-pgpass.html
+As a result, here is the standard upgrade plan for pg_auto_failover:
 
+  1. Put all the current secondary nodes in maintenance to prevent from any
+     failover to happen.
 
-Operations
-----------
+	 Run the follow SQL query on the monitor database to make sure your
+	 setup current state allows for an upgrade operation to happen::
+
+		select nodeid, nodename, nodeport, goalstate, reportedstate
+		  from pgautofailover.node
+		 where reportedstate = 'secondary'
+		   and not exists
+			  (
+			   select nodeid
+				 from pgautofailover.node
+				where goalstate <> reportedstate
+				   or reportedstate not in ('primary', 'secondary')
+			  )
+		order by nodeid;
```

Comment:
It would be nice if we had a command line tool to do this. Something like:
```
pg_autoctl status --is-stable
```

And for the enable/disable maintainance queries we could have:
```
pg_autoctl enable maintainance --all
pg_autoctl disable maintainance --all
```

Or maybe something like:
```
pg_autoctl upgrade prepare
pg_autoctl upgrade finish
```

That seems nicer than telling people to connect manually to the monitor database, which usually they shouldn't do.

This might be better in a separate PR though.

Reply:
Good ideas, even if the exact CLI/API needs to be refined. On a first approach I'm thinking it might make sense to have those operations part of the `pg_autoctl perform` sub-commands, the main (only?) place where we initiate cluster wide operations at the moment.

Let's open a separate issue/PR for this!

Reply:
Thanks to restarting the "node-active" sub-process on extension version mismatch on the monitor, the upgrade procedure can be simplified down to two steps, and none of the preparation queries are needed anymore.

  1. upgrade the OS packages on every node
  2. restart the monitor